### PR TITLE
Feat: Initial implementation of Tabs component

### PR DESCRIPTION
**Pull request by ajainarayanan**
PR for #72 

##### Note:
- Not sure if I adhered to coding style. Please let me know if I missed something
- Will squash all the commits if the changes are ok
- Have added some comments, places where I thought needed some explanation. Please feel free to remove them if you think its not right.
- If someone can point me how to update the test coverage would make sure it is 100%

### CODE REVIEW

**Comment by eddywashere** [comment]
Couple of comments

I think this could use the .tab-content, .tab-pane classes and .active from http://v4-alpha.getbootstrap.com/components/navs/#using-data-attributes (fade option could be part of a different pr). I should have mentioned that in GitHub issue.

Also, instead of iterating through children and setting state, you could use context to simplify the TabContent component https://facebook.github.io/react/docs/context.html.

Ex: add getChildContext to TabContent, return {activeTab: this.props.activeTab}. In TabPane, it can accept a 2nd argument, context, from there you can compare props.tabId to context.activeTab and add the .active class if they are equal.

For coverage, there's a folder in test directory called coverage and somewhere inside is a root index.html. I usually open that in a browser and sort by missing coverage

Thanks for tackling this! feel free to ping me here or in slack if you need anything. The link to it is in contributing.md

**Comment by ajainarayanan** [comment]
@eddywashere Will work on this weekend. Just one minor question, react docs mentions using context only when it is absolutely necessary and try to avoid it as much as possible. Is there any specific reason why we should use contextTypes for communicating with children? The props seemed to be simple. Is there any specific reason we should use contextTypes? (Sorry if my question is a little silly, I am relatively new to react :))

**Comment by eddywashere** [comment]
Glad you brought that up. "Using context makes your components more coupled" I interpret that as a positive effect in this case. TabPanes are coupled to TabContent. It's partially to save time passing the same prop to child components. If the API changes we could always switch to cloning children(TabPanes) with the activeTab prop from TabContent. That wouldn't cause a breaking change and seems like an easy win for keeping things simple.

**Comment by ajainarayanan** [comment]
@eddywashere That was super useful! Have updated the PR. Please review and let me know I have missed anything.

File: src/TabContent.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      activeTab: this.props.activeTab
+    };
+  }
+  getChildContext() {
+    return {
+      activeTabId: this.state.activeTab
+    };
+  }
+  componentWillReceiveProps(nextProps) {
+    this.setState({
```
Comment (eddywashere) [inline src/TabContent.js]: you should check if `nextProps.activeTab !== this.props.activeTab` before using setState

File: src/TabContent.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      activeTab: nextProps.activeTab
+    });
+  }
+  render() {
+    return (
+      <div className={classnames('tab-content', this.props.className)}>
```
Comment (eddywashere) [inline src/TabContent.js]: consider moving `classnames('tab-content', this.props.className)` to a const called classes. It's not across all components but trying to standardize that.

File: src/TabPane.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  className: PropTypes.string,
+  tabId: PropTypes.any
+};
+const contextTypes = {
+  activeTabId: PropTypes.any
+};
+
+export default function TabPane({ className, tabId, children }, context) {
+  return (
+    <div className={classnames('tab-pane', className, { active: tabId === context.activeTabId })}>
```
Comment (eddywashere) [inline src/TabPane.js]: same classes const here

File: src/TabContent.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+const childContextTypes = {
+  activeTabId: PropTypes.any
+};
+
+export default class TabContent extends Component {
+  constructor(props) {
+    super(props);
+    this.childRefs = {};
```
Comment (eddywashere) [inline src/TabContent.js]: this can be removed now

File: src/TabPane.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+const propTypes = {
+  children: PropTypes.node,
+  className: PropTypes.string,
+  tabId: PropTypes.any
+};
+const contextTypes = {
+  activeTabId: PropTypes.any
+};
+
+export default function TabPane({ className, tabId, children }, context) {
```
Comment (eddywashere) [inline src/TabPane.js]: could you move these props into something like the pattern in other components, [example](https://github.com/reactstrap/reactstrap/blob/af874bc9001be3bf692a998714ebe7aa4f3d7bbb/src/Card.js#L18-L26).

This makes it easier to pass through attributes/props that aren't explicitly called out.

File: src/TabContent.js (hunk 1 above)
Comment (ajainarayanan) [inline src/TabContent.js]: Not sure why github is not collapsing or updating this comment but this is fixed

**Comment by eddywashere** [comment]
thanks a ton for working through this @ajainarayanan! I'll squash and merge via github.

File: test/Tabs.spec.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    tab1.find('.nav .nav-item .nav-link').at(1).simulate('click');
+    expect(tab1.find('.nav .nav-item .nav-link').at(0).hasClass('active')).toBe(false);
+    expect(tab1.find('.nav .nav-item .nav-link').at(1).hasClass('active')).toBe(true);
+    expect(tab1.find('.tab-content .tab-pane').at(1).hasClass('active')).toBe(true);
+    tab1.find('.nav .nav-item .nav-link').at(0).simulate('click');
+  });
+  it('should show no active tabs if active tab id is unknown', () => {
+    let tab1 = mount(<TabsExample />);
+    const instance = tab1.instance();
+    expect(instance instanceof TabsExample).toBe(true);
```
Comment (TheSharpieOne) [inline test/Tabs.spec.js]: are you testing the example?

Comment (ajainarayanan) [inline test/Tabs.spec.js]: The TabExample is a wrapper around the Tab. The aim in this test is to have a click initiated from the wrapper that toggles the TabContent (in this case an invalid case). The clicking and managing ActiveTab is not in the scope of the Tab component. The test still focuses on testing the Tab and not the TabsExample though.

Comment (TheSharpieOne) [inline test/Tabs.spec.js]: It's just this except tests the example directly, seems pointless.

**Pull request merged**